tasks/change-granule-collection-s3/src/index.ts

```diff
@@ -0,0 +1,560 @@
+  }
+  return match;
+}
+
+async function cmrFileCollision(
+  sourceFile: ValidApiGranuleFile,
+  targetFile: ValidApiGranuleFile,
+  cmrObject: Object
+) {
+  // if these are the same, we need to *update* that metadata
+  if (objectSourceAndTargetSame(sourceFile, targetFile)) {
+    return false;
+  }
+  if (!await pRetry(
+    () => s3ObjectExists({
+      Bucket: targetFile.bucket,
+      Key: targetFile.key,
+    }),
+    { retries: 5, minTimeout: 2000, maxTimeout: 2000 }
+  )) {
+    return false;
+  }
+  if (!await metadataCollisionsMatch(targetFile, cmrObject)) {
+    throw new DuplicateFile(
+      `metadata file Bucket: ${targetFile.bucket}, Key: ${targetFile.key} already exists.` +
+      'and does not appear to belong to the collection being moved'
+    );
+  }
+  return false;
+}
+
+async function copyFileInS3({
+  sourceFile,
+  targetFile,
+  cmrObject,
+  s3MultipartChunksizeMb,
+}: {
+  sourceFile: ValidApiGranuleFile,
+  targetFile: ValidApiGranuleFile,
+  cmrObject: Object,
+  s3MultipartChunksizeMb?: number,
+}): Promise<void> {
+  if (isCMRMetadataFile(targetFile)) {
+    if (!(await cmrFileCollision(sourceFile, targetFile, cmrObject))) {
+      const metadataString = CMRObjectToString(targetFile, cmrObject);
+      await pRetry(
+        () => uploadCMRFile(targetFile, metadataString),
+        { retries: 5, minTimeout: 2000, maxTimeout: 2000 }
+      );
+    }
+    return;
+  }
+  if (await s3CopyNeeded(sourceFile, targetFile)) {
+    await pRetry(
+      () => copyObject({
+        sourceBucket: sourceFile.bucket,
+        sourceKey: sourceFile.key,
+        destinationBucket: targetFile.bucket,
+        destinationKey: targetFile.key,
+        chunkSize: s3MultipartChunksizeMb,
+      }),
+      { retries: 5, minTimeout: 2000, maxTimeout: 2000 }
+    );
+  }
+}
+/**
+ * Copy granule files in s3.
+ * Any CMRfile will not truly be "copied" but pushed up to new location with new metadata contents
+ * any other file will be checked to be sure a copy is actually needed, and then copied over
+ */
+async function copyGranulesInS3({
+  sourceGranules,
+  targetGranules,
+  cmrObjects,
+  s3MultipartChunksizeMb,
+}: {
+  sourceGranules: Array<ValidGranuleRecord>,
+  targetGranules: Array<ValidGranuleRecord>,
+  cmrObjects: { [granuleId: string]: Object },
+  s3MultipartChunksizeMb?: number,
+}): Promise<void> {
+  const sourceGranulesById = keyBy(sourceGranules, 'granuleId');
+
+  const copyOperations = flatten(await Promise.all(
+    targetGranules.map(
+      async (targetGranule) => {
+        const sourceGranule = sourceGranulesById[targetGranule.granuleId];
+        if (!sourceGranule) {
+          throw new AssertionError({ message: 'no source granule for your target granule by ID' });
+        }
+        if (!sourceGranule.files || !targetGranule.files) {
+          // this needs to be callable for later pmap, but do nothing
+          return noop;
```

Comment:
Why does this need to be a method, versus `undefined`?    The value being returned below is an array of promises...